src/Parsers/ASTSelectQuery.cpp

```diff
@@ -474,4 +476,14 @@ void ASTSelectQuery::setFinal() // NOLINT method can be made const
     tables_element.table_expression->as<ASTTableExpression &>().final = true;
 }
 
+bool ASTSelectQuery::hasQueryParameters() const
+{
+    if (!has_query_parameters.has_value())
+    {
+        has_query_parameters = !analyzeReceiveQueryParams(std::make_shared<ASTSelectQuery>(*this)).empty();
```

Comment:
I think its better to keep input `ASTPtr`, as this can be used by other queries too in future. Unable to use shared_from_this() here as it gives `shared_ptr<const DB::IAST>` as this is const function and `ASTPtr` is `shared_ptr<DB::IAST>`, maybe that is why we dont use it in `clone` too.